### PR TITLE
psudohash: init at unstable-2023-05-15

### PR DESCRIPTION
**Pull request by exploitoverload**
###### Description of changes

Adds psudohash - a password list generator for orchestrating brute force attacks and cracking hashes.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Related to #81418

### CODE REVIEW

**Comment by NickCao** [comment]
It looks like a simple python script with no external dependencies, we can just use `stdenv.mkDerivation` and `patchShebangs`and call it a day.

**Comment by exploitoverload** [comment]
> It looks like a simple python script with no external dependencies, we can just use `stdenv.mkDerivation` and `patchShebangs`and call it a day.

@NickCao Done! Updated with patchShebangs and stdenv :wink:

File: pkgs/tools/security/psudohash/default.nix [hunk 1]
```diff
@@ -0,0 +1,46 @@
+{
+  lib,
+  fetchFromGitHub,
+  makeWrapper,
```
Comment (NickCao) [inline pkgs/tools/security/psudohash/default.nix:4]: Drop the unused `makeWrapper` argument.

File: pkgs/tools/security/psudohash/default.nix [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  version = "unstable-2023-05-15";
+
+  src = fetchFromGitHub {
+    owner = "t3l3machus";
+    repo = "psudohash";
+    rev = "2d586dec8b5836546ae54b924eb59952a7ee393c";
+    hash = "sha256-l/Rp9405Wf6vh85PFrRTtTLJE7GPODowseNqEw42J18=";
+  };
+
+  nativeBuildInputs = [ python3 ];
```
Comment (NickCao) [inline pkgs/tools/security/psudohash/default.nix:20]: Should be `buildInputs` as it is used at runtime not build time.

File: pkgs/tools/security/psudohash/default.nix [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  dontBuild = true;
```
Comment (NickCao) [inline pkgs/tools/security/psudohash/default.nix:21]: This line can be dropped as there is no makefile and stdenv won't attempt a build.

File: pkgs/tools/security/psudohash/default.nix [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/bin $out/share/${pname}
+
+    cp -r . $out/share/${pname}
```
Comment (NickCao) [inline pkgs/tools/security/psudohash/default.nix:28]:
```suggestion
```

Do we really need to install the screenshots?

File: pkgs/tools/security/psudohash/default.nix [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+    install -Dm0755 $out/share/${pname}/psudohash.py $out/bin/psudohash
+
+    patchShebangs $out/bin/psudohash
```
Comment (NickCao) [inline pkgs/tools/security/psudohash/default.nix:32]: stdenv automatically patches shebangs in installed scripts, this line can also be dropped.

File: pkgs/tools/security/psudohash/default.nix [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  dontBuild = true;
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/bin $out/share/${pname}
+
+    cp -r . $out/share/${pname}
+
+    install -Dm0755 $out/share/${pname}/psudohash.py $out/bin/psudohash
```
Comment (NickCao) [inline pkgs/tools/security/psudohash/default.nix:30]:
```suggestion
    install -Dm555 psudohash.py $out/bin/psudohash
    install -Dm444 common_padding_values.txt $out/share/psudohash/common_padding_values.txt
```

Comment (exploitoverload) [reply pkgs/tools/security/psudohash/default.nix]: @NickCao Thanks for the comments! I've just pushed the changes, you were right, the screenshots weren't necessary :joy:

**Review by NickCao — APPROVED** [review]

**Pull request merged**